libsolidity/formal/UniqueErrorReporter.h

```diff
@@ -0,0 +1,55 @@
+#include <liblangutil/ErrorReporter.h>
+
+namespace solidity::frontend
+{
+
+struct UniqueErrorReporter
+{
+	UniqueErrorReporter(): errorReporter(uniqueErrors) {}
+
+	void warning(langutil::ErrorId _error, langutil::SourceLocation const& _location, std::string const& _description)
+	{
+		if (!has(_error, _location, _description))
+			errorReporter.warning(_error, _location, _description);
+		set(_error, _location, _description);
+	}
+
+	void warning(
+			langutil::ErrorId _error,
+			langutil::SourceLocation const& _location,
+			std::string const& _description,
+			langutil::SecondarySourceLocation const& _secondaryLocation
+			)
+	{
+		if (!has(_error, _location, _description))
+			errorReporter.warning(_error, _location, _description, _secondaryLocation);
+		set(_error, _location, _description);
+	}
+
+	void warning(langutil::ErrorId _error, std::string const& _description)
+	{
+		if (!has(_error, {}, _description))
+			errorReporter.warning(_error, _description);
+		set(_error, {}, _description);
+	}
+
+	bool has(langutil::ErrorId _error, langutil::SourceLocation const& _location, std::string const& _description)
+	{
+		return seenErrors.count({_error, _location, _description});
+	}
+
+	void set(langutil::ErrorId _error, langutil::SourceLocation const& _location, std::string const& _description)
```

Comment:
I think that the name `set` does not make it clear enough whether it also reports the error or not.
```suggestion
	void markAsSeen(langutil::ErrorId _error, langutil::SourceLocation const& _location, std::string const& _description)
```
I'd also rename `has()` to `seen()`.